module/zcommon/zfs_fletcher.c

```diff
@@ -501,7 +508,13 @@ fletcher_4_byteswap_impl(const void *buf, uint64_t size, zio_cksum_t *zcp)
 	const fletcher_4_ops_t *ops = fletcher_4_impl_get();
 
 	ops->init_byteswap(&ctx);
+	if (ops->uses_fpu == B_TRUE) {
+		kfpu_begin();
+	}
 	ops->compute_byteswap(&ctx, buf, size);
+	if (ops->uses_fpu == B_TRUE) {
+		kfpu_end();
+	}
```

Comment:
`kfpu_begin()` should be called before `init_byteswap()` and `kfpu_end()` should be called after `fini_byteswap()`. That way the init and fini functions are allowed to use vector instructions. A future patch to compile GNU C Vector code likely will make use of this.

The same applies to the native versions.

Reply:
The intention was to minimize the time we're running with interrupts and preemption disabled. So this is a trade-off between ease of future changes and an actual  advantage. I'm fine either way, let's see what other reviewers think.

Reply:
59493b63c18ea223857066218d6a58b67eb88159 already made a decent reduction in the total time that we spend with interrupts disabled not that long ago by not doing context save/restore multiple times per ABD. At the same time, it did `kfpu_begin()` before the rest of the init function and `kfpu_end()` after the rest of the fini function, which passed review.

An extra microsecond or two with interrupts disabled will not matter in the slightest since the systems we support do not give real time guarantees. If we were running on a system that tried to give real time guarantees and it were so pedantic to give guarantees that were accurate to a level where a microsecond or two difference would matter, we probably would not be allowed to run with interrupts disabled at all, since disabling interrupts to use SIMD could cause the real time guarantee to be violated. I am not sure if even real-time Linux is that strict.

Anyway, what is important here is not minimizing time with interrupts disabled, but minimizing the total time spent doing processing. If the interrupts are disabled slightly longer than absolutely necessary to reduce the total CPU time spent, then that is preferable. In this case, doing it this way is mostly to prepare for a future small reduction in CPU time. Since we know that there is a possibility of that here, I would prefer us to do it that way so that we do not need another patch. We really are not gaining anything from an extra microsecond or two with interrupts enabled per checksum operation, so there is no need to be frugal on that level.

Reply:
Done.